Schedule2 (alpha2)

saerdnaer opened the pull request:
breaking changes to the schedule.json schema, thereby a own new version called `schedule2`, see https://github.com/voc/schedule/commit/6632bbbcd79411592c464f48212dcd434d54371c for an comparison between the formats
* some unused fields get dropped, e.g. `public_name`, `daysCount`, `timeslot_duration`, …
* Track: `color` -> `colors` to allow multiple color values e.g. for light an dark – and to be uniform to same attribute on Conference
* Track: gets a `guid` field
* translations of abstract, description etc. get switched from `Accept` header `"string"` to object  `{"en": "string", "de": "string"}`, but the old way is still allowed
* Day: `day_start` and `day_end` get's unified to `start` and `end`
* events and persons get it's own top level field, and are no longer nested
  * but events still have to be sorted by date and room.order as before by the schedule generator/serializer
   * connection from Event to Person is named `Participant` and has a `role` attribute 
* Images like `logo` and `avatar` are now an object `{"url": "uri-string"}` instead of `"uri-string"`
* Event has now `start` and `end` instead of `date` and `duration`, fixing #129 

TODOs
* [x] decide how the references on Event to room should look like
* [x] decide how the references on Event to track should look like
* [x] decide if `Participant.role` should be an enum
* [x] add `Reference.service` enum based on ucdata
* [x] add `languages: ['en', 'de']` on `Conference`
* [x] add `defaults: { theme: 'dark', language: 'en' }`… on `Conference`?
* [x] should we change all enum values to UPPER_CASE?
* [x] optional: expose state of event, so clients can show that a Event was canceled
* [ ] optional: add support for full-day events e.g. soldering/climbing workshops
* [ ] optional: expose special events like break or relay/streaming of a talk into a secondary room

Optional TODOs (for the future?):
* [ ] optional: expose live information like 
   * [ ] operational delays: real start, expected delay in minutes as integers
   * [ ] room occupancy: parts of capacity
   * [ ] favourite counts
   * [ ] …

saerdnaer committed:
chore: add excerpt of 38C3 schedule.json generated by the hub

Signed-off-by: Andreas Hubel <andreas.hubel@pub.tech>

saerdnaer committed:
chore: add excerpt of 38C3 schedule.json generated by the hub

Signed-off-by: Andreas Hubel <andreas.hubel@pub.tech>

saerdnaer committed:
first draft for schedule2 json schema + example

Signed-off-by: Andreas Hubel <andreas.hubel@pub.tech>

saerdnaer committed:
first draft for schedule2 json schema + example

Signed-off-by: Andreas Hubel <andreas.hubel@pub.tech>

saerdnaer commented:
@toto  Sollen wir die kompletten Services hierher in schedule2 mit übernehmen?

https://github.com/toto/conference-import/blob/e51e5a35788a6735a724ece59e84018efd18dc41/src/models/link.ts#L1-L8

toto commented:
Not all pick: that's historical baggage. How about
- Video: YouTube, Vimeo 
- Social: Facebook, ActivityPub, LinkedIn, GitHub and threads. 
Plus "web" as the default.

johnjohndoe commented on validator/json/schema2.json:
```diff
@@ -0,0 +1,468 @@
+          "type": ["string", "null", "object"]
+        },
+        "description": {
+          "type": ["string", "null", "object"]
```
Also, track names could be translated.

saerdnaer replied:
@johnjohndoe regarding tracks: I asked the content team and we came to the conclusion that tracks are similar to person names aka they stay the same independent of the language

johnjohndoe commented on validator/json/examples/hub-schedule2.json:
```diff
@@ -0,0 +1,324 @@
+        "slug": "38c3-38c3-opening-ceremony",
+        "title": "38C3: Opening Ceremony",
+        "subtitle": null,
+        "language": "en",
```
Should we rename `language` to `languages`? It is not limited to one language if I remember correctly.

Kunsi replied:
Depends if we want to include the languages of possible translations. If this is the language of the talk itself, there can only ever be one i think.

johnjohndoe replied:
I believe talks can still be hold in multiple languages, e.g. where speaker1 talks English and speaker2 talks German.

toto replied:
Let’s have one language the talk is held in. If we want translations that should be a different field / indicated in some way.

johnjohndoe replied:
I did not refer to translations here. I really meant a talk which is hold in different languages and is therefore announced as such, e.g. `en, de`.

rixx replied:
Is there any system that currently supports this? AFAIK neither frab nor pretalx do; pretalx certainly assumes that a talk is in one language.

johnjohndoe replied:
I remember seeing it in the Android app at one point. I would need to investigate to find out what the source backend was.

johnjohndoe replied:
I could be that session with multiple languages originate from a wiki import at 38C3.

saerdnaer replied:
for self-organized sessions created in the Hub, there is the option to select a ordered list of possible languages, via a dropdown. In general I would see this field as main/original language and add an additional translations field.

johnjohndoe replied:
What does the UI for SOS look like / tell about the language selection? Are they being selected as the speakers languages or as the available translations?

johnjohndoe commented on validator/json/examples/hub-schedule2.json:
```diff
@@ -0,0 +1,324 @@
+        "start": "2024-12-27T10:30:00+01:00",
+        "end": "2024-12-27T11:00:00+01:00",
+        "duration": "00:30",
+        "room": "Saal 1",
```
Should we rename `room` to `roomName` to clarify its purpose?

Kunsi replied:
`room_name` would be consistent with the other names, but i agree with that proposal.

toto replied:
Yes 👍 but please let’s add `room_guid`. Just matching via name otherwise.

johnjohndoe commented on validator/json/examples/hub-schedule2.json:
```diff
@@ -0,0 +1,324 @@
+        {
+          "guid": "3dc5daf5-ba21-4577-97e4-fb6e4b5d40e1",
+          "name": "Art & Beauty",
+          "colors": {
```
Not sure if a preferred color scheme property is needed. Apps typically follow either the system settings or provide a way for users to customize the scheme within the app as show in the screenshot:

![theme](https://github.com/user-attachments/assets/1cdf01d5-a868-45fd-b9cc-d6470ee72d90)

Kunsi replied:
Keep in mind we don't always have that. For example, pretalx always uses light mode as of now, so that would be preferred. info-beamer in itself does also not know what the user preference is, so it would use the organizer preferences.

johnjohndoe replied:
Thank you for the clarification. @saerdnaer It would be great to provide some context in the issue description up there and list all the clients which consume the JSON.

Kunsi replied:
I don't believe we have a list of all the apps that might consume the json schedule.

johnjohndoe replied:
It does not have to be complete to be valuable.

johnjohndoe commented on validator/json/examples/hub-schedule2.json:
```diff
@@ -0,0 +1,324 @@
+          "type": "lecturehall",
+          "stream_id": "s1",
+          "capacity": 3025,
+          "description": {
```
For the language, I would also follow the system settings resp. the app specific language settings.

johnjohndoe replied:
If an app request the description in a language which is **not present** then the backend could return the description in a **fallback language**. This can be **transparent** to the app.

Kunsi replied:
Same as with colours, we don't always have that information. Information screeens will default to the default locale unless explicitely set differently.

johnjohndoe commented on validator/json/examples/hub-schedule2.json:
```diff
@@ -0,0 +1,324 @@
+        "guid": "270fa44c-821a-577f-8ca3-33e8d4a8c2a7",
+        "start": "2024-12-27T10:30:00+01:00",
+        "end": "2024-12-27T11:00:00+01:00",
+        "duration": "00:30",
```
To be removed.

Kunsi commented on validator/json/examples/hub-schedule2.json:
```diff
@@ -0,0 +1,324 @@
+        ],
+        "links": [
+          {
+            "url": "https://domain.tld/",
```
Just noticed that, we should use `example.{com,org,net}` here.

saerdnaer committed:
chore: suggestions from review

saerdnaer committed:
chore: suggestions from review

saerdnaer committed:
chore: drop duration from hub-schedule2.json example file

saerdnaer committed:
chore: drop duration from hub-schedule2.json example file

saerdnaer committed:
feat(schedule2): add Language type definition

saerdnaer committed:
feat(schedule2): add Language type definition

saerdnaer committed:
feat(schedule2): change Event.room to RoomReference object

saerdnaer committed:
feat(schedule2): change Event.room to RoomReference object

saerdnaer committed:
feat(schedule2): change Event.track to TrackReference object

saerdnaer committed:
feat(schedule2): change Event.track to TrackReference object

saerdnaer committed:
chore(schedule2): extract slug pattern as own type defintion

saerdnaer committed:
chore(schedule2): extract slug pattern as own type defintion

saerdnaer committed:
chore(schedule2): specify optional EventParticipant.role as enum

saerdnaer committed:
chore(schedule2): specify optional EventParticipant.role as enum

saerdnaer committed:
chore(schedule2): cleanup

saerdnaer committed:
chore(schedule2): cleanup

saerdnaer committed:
feat(schedule2): add meta events for breaks e.g. lunch, dinner, setup etc

saerdnaer committed:
feat(schedule2): add meta events for breaks e.g. lunch, dinner, setup etc

saerdnaer commented on validator/json/examples/hub-schedule2.json:
```diff
@@ -213,7 +213,9 @@
-        "track": "CCC",
+        "track": {
+          "slug": "ccc"
+        },
```
I changed the track reference, to be uniform with the room refeference

johnjohndoe replied:
Are we missing a reference to the `guid` of the track here?

saerdnaer replied:
I am not yet at the point where I see a track guid as really required, a persistent slug is a good enough "ID" for me for this kind of entity... Otherwise we have to discuss if they have the same guid across each conference instance etc.

saerdnaer committed:
feat(schedule2): add service enum on Reference type definition

saerdnaer committed:
feat(schedule2): add service enum on Reference type definition

saerdnaer committed:
feat(schedule2): add feedback + origin on Reference.type

saerdnaer committed:
feat(schedule2): add feedback + origin on Reference.type

toto commented:
Absolutely canceled should be in there at least. Not sure about the rest.

saerdnaer committed:
feat(schedule2): add defaults object to Conference

saerdnaer committed:
feat(schedule2): convert all enum values to UPPER_CASE

Standardize enum strings across schema2.json and example file to use UPPER_CASE/SCREAMING_SNAKE_CASE convention.

- Event types: talk → TALK, workshop → WORKSHOP, etc.
- Person roles: speaker → SPEAKER, moderator → MODERATOR, etc.
- Meta event types: break → BREAK, meta → META
- Room types: lecturehall → LECTURE_HALL, stage → STAGE, bbb removed, online → VIRTUAL
- Recording features: record_by_default → RECORD_BY_DEFAULT, etc.
- Reference/attachment types: slides → SLIDES, feedback → FEEDBACK, origin → ORIGIN, etc.

saerdnaer committed:
feat(schedule2): rename assembly to parent, generalize room grouping

Rename HubAssembly → RoomGrouping with added `type` field (Assembly, Building, Floor, Area, Other). Room's `assembly` field renamed to `parent`, now accepts RoomReference or RoomGrouping. Update example accordingly.

saerdnaer committed:
feat(schedule2): add type field to Room and Image objects

saerdnaer committed:
chore: cleanup

saerdnaer committed:
set version to alpha-2

saerdnaer committed:
feat(schema2): add state to Event

Tracks lifecycle state of events (DRAFT → SUBMITTED → ACCEPTED/REJECTED → CONFIRMED, plus DELAYED/CANCELED/WITHDRAWN/DELETED). Defaults to CONFIRMED.

saerdnaer commented:
Releasing alpha-Version of schedule2, moving the remaining optional task to own issues/PRs.
